### PR TITLE
Spark DataFrames handled as a type if using spark

### PR DESCRIPTION
**Pull request by kumare3**
- This is also an overhaul of the schema system to support remote IO
   based dataframes and regular pandas style dataframes. All of them can
be added at the user layer by implementing 2 classes SchemaReader
/ SchemaWriter.
 - If the new dataframe is to be added as a supported type then, it
   contributor needs to implement the TypeTransformer interface as well

### CODE REVIEW

File: flytekit/types/schema.py [hunk 1]
```diff
@@ -132,17 +259,31 @@ def open(
...
-            self._downloader(self._remote_path, self._local_path)
-            self._downloaded = True
-        return SchemaEngine.open(t=dataframe_fmt, s=self, mode=override_mode if override_mode else self._supported_mode)
+        mode = override_mode if override_mode else self._supported_mode
+        h = SchemaEngine.get_handler(dataframe_fmt)
+        if not h.handles_remote_io:
+            # The Schema Handler does not manage its own IO, and this it will expect the files are on local file-system
+            if self._supported_mode == SchemaOpenMode.READ and not self._downloaded:
+                # Only for readable objects if they are not downloaded already, we should download them
+                # WRite objects should already have everything written to
```
Comment (wild-endeavor) [inline flytekit/types/schema.py:276]:
```suggestion
                # Write objects should already have everything written to
```

File: tests/flytekit/unit/annotated/test_type_hints.py [hunk 2]
```diff
@@ -1104,3 +1105,53 @@ def wf(x: int, y: int) -> Datum:
...
+        new_df = session.createDataFrame([("Bob", 10)], my_schema.column_names())
+        return df.union(new_df)
+
+    @workflow
+    def my_wf() -> my_schema:
+        df = my_dataset()
+        return my_spark(df=df)
+
+    x = my_wf()
+    reader = x.open(pandas.DataFrame)
```
Comment (wild-endeavor) [inline tests/flytekit/unit/annotated/test_type_hints.py:1152]: what happens if `pandas.DataFrame` is not specified?

**Review by wild-endeavor — APPROVED** [review]
some minor comments

File: tests/flytekit/unit/annotated/test_type_hints.py (hunk 2 above)
Comment (kumare3) [reply tests/flytekit/unit/annotated/test_type_hints.py:1152]: default is pandas.DataFrame actually, I can drop it

**Pull request merged**